src/lib/visualization/cardinality_writer.cpp

```diff
@@ -0,0 +1,218 @@
+    const auto lqp_description = _lqp_nodes[node]->description();
+    const auto pqp_description = _pqp_nodes[node]->description();
+
+    // Get estimated and actual output cardinalities (for next columns)
+    const auto lqp_row_count = _cardinality_estimator.estimate_cardinality(_lqp_nodes[node]);
+
+    auto pqp_row_count = NAN;
+    const auto& performance_data = *_pqp_nodes[node]->performance_data;
+
+    if (_pqp_nodes[node]->executed() && performance_data.has_output) {
+      pqp_row_count = static_cast<float>(performance_data.output_row_count);
+    }
+
+    // Write benchmark id, node descriptions and cardinalities in CSV
+    // Use semicolon as seperator instead of comma to be able to work with database values containing commas
+    output_estimations << benchmark_item_id << ";" << lqp_description << ";" << pqp_description << ";" << lqp_row_count
+                       << ";" << pqp_row_count << "\n";
+  }
+
+  // Write debug information about possible different numbers of lqp and pqp nodes
+  output_debug << benchmark_item_id << ";" << lqp_nodes_size << ";" << pqp_nodes_size << "\n";
+
+  // Close file handles
+  output_estimations.close();
+  output_debug.close();
+}
+
+void CardinalityWriter::_build_lqp_graph(const std::vector<std::shared_ptr<AbstractLQPNode>>& lqp_roots) {
+  std::unordered_set<std::shared_ptr<const AbstractLQPNode>> visited_nodes;
+  ExpressionUnorderedSet visited_sub_queries;
+
+  for (const auto& root : lqp_roots) {
+    _build_lqp_subtree(root, visited_nodes, visited_sub_queries);
+  }
+}
+
+void CardinalityWriter::_build_lqp_subtree(const std::shared_ptr<AbstractLQPNode>& node,
+                                           std::unordered_set<std::shared_ptr<const AbstractLQPNode>>& visited_nodes,
+                                           ExpressionUnorderedSet& visited_sub_queries) {
+  if (visited_nodes.find(node) != visited_nodes.end()) return;
+  visited_nodes.insert(node);
+
+  // Add current node to our container
+  _lqp_nodes.push_back(node);
+
+  if (node->left_input()) {
+    auto left_input = node->left_input();
+    _build_lqp_subtree(left_input, visited_nodes, visited_sub_queries);
+  }
+
+  if (node->right_input()) {
+    auto right_input = node->right_input();
+    _build_lqp_subtree(right_input, visited_nodes, visited_sub_queries);
+  }
+
+  // Traverse subqueries
+  for (const auto& expression : node->node_expressions) {
+    visit_expression(expression, [&](const auto& sub_expression) {
+      const auto subquery_expression = std::dynamic_pointer_cast<LQPSubqueryExpression>(sub_expression);
+      if (!subquery_expression) return ExpressionVisitation::VisitArguments;
+
+      if (!visited_sub_queries.emplace(subquery_expression).second) return ExpressionVisitation::VisitArguments;
+
+      _build_lqp_subtree(subquery_expression->lqp, visited_nodes, visited_sub_queries);
+
+      return ExpressionVisitation::VisitArguments;
+    });
+  }
+}
```

Comment:
Is that part obtained from another file, i.e., the same functionality used multiple times?

Reply:
Almost. The functionality is similar, but different in a few key places. We obtained it from the `lqp_visualizer.cpp` and `pqp_visualizer.cpp` respectively, but changed e.g. the traversion order for the `pqp` and removed a lot of actions that are unnecessary for only printing out cardinality estimations. The common part / the part we reused is the `lqp` and `pqp`-specific traversion of the lqp and `pqp` trees. To refactor it we probably would have to define a new class and then pass around a lot of lambdas to change the behavior during the traversion, which would perhaps make it harder to understand the code.

Reply:
That's fine, you don't have to do that.